### PR TITLE
Fixes a couple of edge cases with sum/reduction WGSL layer.

### PR DESCRIPTION
**Pull request by unrenormalizable**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

- Recreating the PR #1231 as I messed up my fork
- The changes are issues identified during #1173

### Changes

Fixes a couple of edge cases with sum/reduction WGSL layer.

### Testing

- Added a new test case
- run-checks all

### CODE REVIEW

File: crates/burn-wgpu/src/template/reduction/recursive_sum.wgsl [hunk 1]
```diff
@@ -21,14 +22,23 @@ fn main(
...
     workgroupBarrier();
 
+    var num_data = WORKGROUP_SIZE;
+    if (id_local == last_index) {
+        num_data = min(WORKGROUP_SIZE, arrayLength(&input) % WORKGROUP_SIZE);
+        if (num_data == 0u) {
+            num_data = WORKGROUP_SIZE;
+        }
+    }
+
```
Comment (nathanielsimard) [inline crates/burn-wgpu/src/template/reduction/recursive_sum.wgsl:38]: We are going to remove that kernel pretty soon and replace it with the same ones as the reduce_dim (two version).

**Review by nathanielsimard — COMMENTED** [review]
We are moving away from the templating API, and we will rewrite all WGSL kernels in `burn-wgpu`, so I'll close this PR since this part of the codebase will be deleted soon.

Hopefully, you didn't spend much time on this 😅

File: crates/burn-wgpu/src/template/reduction/recursive_sum.wgsl (hunk 1 above)
Comment (unrenormalizable) [reply crates/burn-wgpu/src/template/reduction/recursive_sum.wgsl:38]: no worries at all. helped me ramp up on the codebase. thanks for looking into it.
